src/SourceBuild/content/test/Microsoft.DotNet.SourceBuild.SmokeTests/ExclusionsHelper.cs

```diff
@@ -0,0 +1,179 @@
+            })
+            .SelectMany(parts =>
+                // Create a new object for each suffix
+                parts.Suffixes.Select(suffix => new
+                {
+                    parts.Line,
+                    Suffix = suffix
+                })
+            )
+            .GroupBy(
+                parts => parts.Suffix,
+                parts => parts.Line
+            )
+            .ToDictionary(
+                group => group.Key,
+                group => new HashSet<string>(group)
+            );
+    }
+
+    private static void RemoveUsedExclusion(string exclusionsFileName, string exclusion, string suffix)
+    {
+        if (FileToSuffixToUnusedExclusionsMap.TryGetValue(exclusionsFileName, out Dictionary<string, HashSet<string>> suffixToExclusions)
+            && suffixToExclusions.TryGetValue(suffix, out HashSet<string> exclusions))
+        {
+            exclusions.Remove(exclusion);
+        }
+    }
+
+    private static void RemoveUnusedExclusionsFromBaselines()
+    {
+
+        foreach (KeyValuePair<string, Dictionary<string, HashSet<string>>> fileToUnusedExclusions in FileToSuffixToUnusedExclusionsMap)
+        {
+            string exclusionsFileName = fileToUnusedExclusions.Key;
+            string exclusionsFilePath = Path.Combine(BaselineHelper.GetAssetsDirectory(), exclusionsFileName);
+            var suffixesToUnusedExclusions = fileToUnusedExclusions.Value;
+
+            string[] lines = File.ReadAllLines(exclusionsFilePath);
+
+            var newLines = lines
+                .Select(line =>
+                {
+                    if (suffixesToUnusedExclusions.Values.All(exclusions => exclusions.All(exclusion => !line.Contains(exclusion))))
+                    {
+                        // Line does not contain an unused exclusion, so we can keep it as is
+                        return line;
+                    }
+
+                    string exclusion = line.Split('|')[0];
+                    var unusedSuffixes = suffixesToUnusedExclusions.Where(pair => pair.Value.Contains(exclusion)).Select(pair => pair.Key).ToList();
+
+                    if (unusedSuffixes.Count == 1 && unusedSuffixes[0] == NullSuffix)
+                    {
+                        // Line does not contain any suffixes, so we can remove it entirely
+                        return null;
+                    }
+
+                    string suffixString = line.Split('|')[1].Split('#')[0];
+                    var originalSuffixes = suffixString.Split(',').Select(suffix => suffix.Trim()).ToList();
+                    var newSuffixes = originalSuffixes.Except(unusedSuffixes).ToList();
+
+                    if (newSuffixes.Count == 0)
+                    {
+                        // All suffixes were unused, so we can remove the line entirely
+                        return null;
+                    }
+
+                    return line.Replace(suffixString, string.Join(",", newSuffixes));
```

Comment:
This is a good-size chunk of code and would be better to encapsulate into a method. I try to avoid having too much content in lambda expressions.